logging/nxscope/nxscope_pser.c

```diff
@@ -0,0 +1,297 @@
+  .footlen       = NXSCOPE_CRC_LEN,
+  .hdrlen        = NXSCOPE_HDR_LEN
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxscope_hdr_fill
+ ****************************************************************************/
+
+static void nxscope_hdr_fill(FAR uint8_t *buff, uint8_t id, uint16_t len)
+{
+  FAR struct nxscope_hdr_s *hdr  = NULL;
+
+  hdr = (FAR struct nxscope_hdr_s *) buff;
+
+  hdr->sof = NXSCOPE_HDR_SOF;
+
+  /* Length always as little endian */
+
+#ifdef CONFIG_ENDIAN_BIG
+  hdr->len = bswap_16(len);
+#else
+  hdr->len = len;
+#endif
+
+  hdr->id  = id;
+}
+
+/****************************************************************************
+ * Name: nxscope_frame_get
+ ****************************************************************************/
+
+static int nxscope_frame_get(FAR struct nxscope_proto_s *p,
+                             FAR uint8_t *buff, size_t len,
+                             FAR struct nxscope_frame_s *frame)
+{
+  FAR struct nxscope_hdr_s *hdr = NULL;
+  uint16_t                  crc = 0;
+  int                       ret = OK;
+  int                       i   = 0;
+
+  DEBUGASSERT(p);
+  DEBUGASSERT(buff);
+  DEBUGASSERT(frame);
+
+  UNUSED(p);
+
+  /* Find SOF */
+
+  for (i = 0 ; i < len - NXSCOPE_HDR_LEN; i += 1)
+    {
+      hdr = (FAR struct nxscope_hdr_s *) &buff[i];
+
+      /* Verify SOF */
+
+      if (hdr->sof == NXSCOPE_HDR_SOF)
+        {
+          break;
+        }
+    }
+
+  if (hdr->sof != NXSCOPE_HDR_SOF)
+    {
+      ret = -EINVAL;
+      goto errout;
+    }
+
+  /* Verify len - always little-endian */
+
+#ifdef CONFIG_ENDIAN_BIG
+  hdr->len = bswap_16(hdr->len);
+#endif
+
+  if ((len - i) < hdr->len)
+    {
+      ret = -EINVAL;
+      goto errout;
+    }
+
+  /* Verify crc16 for the whole frame */
+
+  crc = crc16(&buff[i], hdr->len);
+  if (crc != 0)
+    {
+      _err("ERROR: invalid crc16 %d\n", crc);
+      ret = -EINVAL;
+      goto errout;
+    }
+
+  /* Return the frame data */
+
+  frame->id   = hdr->id;
+  frame->data = &buff[i + NXSCOPE_DATA_START];
+  frame->drop = hdr->len + i;
+  frame->dlen = hdr->len - NXSCOPE_BASEFRAME_LEN;
+
+errout:
+  return ret;
+}
+
+/****************************************************************************
+ * Name: nxscope_frame_final
+ ****************************************************************************/
+
+static int nxscope_frame_final(FAR struct nxscope_proto_s *p,
+                               uint8_t id,
+                               FAR uint8_t *buff, FAR size_t *len)
+{
+  uint16_t crc = 0;
+  int      ret = OK;
+
+  DEBUGASSERT(p);
+  DEBUGASSERT(buff);
+  DEBUGASSERT(len);
+
+  if (*len <= NXSCOPE_HDR_LEN)
+    {
+      /* No data */
+
+      ret = -ENODATA;
+      goto errout;
```

Comment:
Just an idea here and other similar places:
just `return -ENODATA;` and remove `goto` and `ret`.

Reply:
I prefer one `return` statement when possible